km/km_cpu_init.c

```diff
@@ -304,12 +306,15 @@ km_vcpu_t* km_vcpu_get(void)
    if (km_vcpu_init(vcpu) != 0) {
       km_mutex_unlock(&machine.vm_vcpu_mtx);
       km_warnx("VCPU init failed");
-      km_vcpu_fini(vcpu, 0);
+      km_vcpu_fini(vcpu);
       return NULL;
    }
    machine.vm_vcpu_cnt++;
    machine.vm_vcpu_run_cnt++;
+   km_lock_vcpu_thr(vcpu);
```

Comment:
Entry and exit to free list is protected by vm_vcpu_mtx why do we need additional lock here,
If we want to protect state transitions on state then do we need to protect all changes to state variable?
Is there a implied  ownership of state variable when vcpu is changing it's own state.

Reply:
Yes, this lock protects this vcpu state transitions.

Reply:
I don't think holding the vcpu_thr lock here does much.
Until we insert the vcpu into machine.vm_vcpus[] nothing will be able to access the vcpu except this function.